In [12]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

# Load the dataset
data = pd.read_csv("Dataset - Factors Influencing Technology Adoption in Consumer Households.csv")

# Preprocessing features from dataset
selected_features = [
    'Age', 'HouseholdIncome', 'Frequency_LaptopDesktop',
    'Frequency_Smartphone', 'Frequency_Tablet', 'Frequency_SmartHomeDevices'
]

# Polynomial Multiple Regression Task
X_poly = data[selected_features].values
y_poly = data['LikelinessToRecommend'].values  # Target variable for regression

data['LikelinessToAdopt'] = data['LikelinessToAdopt'].apply(lambda x: 1 if x >= 3 else 0)
X_class = data[selected_features].values
y_class = data['LikelinessToAdopt'].values  # Target variable for classification

# Split the data into training and testing sets
X_poly_train, X_poly_test, y_poly_train, y_poly_test = train_test_split(X_poly, y_poly, test_size=0.2, random_state=42)
X_class_train, X_class_test, y_class_train, y_class_test = train_test_split(X_class, y_class, test_size=0.2, random_state=42)

# Feature Scaling
scaler = StandardScaler()
X_poly_train_scaled = scaler.fit_transform(X_poly_train)
X_poly_test_scaled = scaler.transform(X_poly_test)

X_class_train_scaled = scaler.fit_transform(X_class_train)
X_class_test_scaled = scaler.transform(X_class_test)

# Display data
data

Age  Gender        EducationLevel  HouseholdIncome  \
0     25  Female     Bachelor's Degree        1500000.0   
1     48  Female     Bachelor's Degree         800000.0   
2     49    Male       Master's Degree       30000000.0   
3     38    Male     Bachelor's Degree          20000.0   
4     45    Male     Bachelor's Degree        5000000.0   
..   ...     ...                   ...              ...   
102   38  Female       Master's Degree         400000.0   
103   25    Male     Bachelor's Degree         840000.0   
104   46    Male       Master's Degree       10000000.0   
105   58  Female  High School Graduate       11200000.0   
106   24  Female     Bachelor's Degree         456000.0   

     Frequency_LaptopDesktop  Frequency_Smartphone  Frequency_Tablet  \
0                          4                     5                 0   
1                          3                     5                 5   
2                          4                     4                 0   
3                          1                     0                 1   
4                          5                     5                 3   
..                       ...                   ...               ...   
102                        5                     5                 3   
103                        4                     5                 3   
104                        4                     5                 3   
105                        5                     5                 0   
106                        1                     5                 0   

     Frequency_SmartHomeDevices  Frequency_WearableDevices  \
0                             4                          0   
1                             4                          5   
2                             2                          0   
3                             1                          1   
4                             5                          5   
..                          ...                        ...   
102                           3                          4   
103                           0                          2   
104                           5                          0   
105                           0                          5   
106                           3                          0   

     Frequency_GamingConsoles  ...  Influence_PromoActivities  \
0                           0  ...                          3   
1                           2  ...                          4   
2                           0  ...                          3   
3                           2  ...                          2   
4                           0  ...                          3   
..                        ...  ...                        ...   
102                         3  ...                          3   
103                         0  ...                          3   
104                         0  ...                          5   
105                         0  ...                          1   
106                         0  ...                          3   

     Influence_SocialPopularity  Influence_ExpertRecommendation  \
0                             5                               4   
1                             4                               4   
2                             3                               3   
3                             3                               3   
4                             3                               3   
..                          ...                             ...   
102                           4                               3   
103                           3                               4   
104                           4                               1   
105                           1                               1   
106                           3                               3   

     Influence_Endorsements  Barrier_HighPrice  Barrier_TechKnowledge  \
0                         2

In [2]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

# Define the model
model_tf = Sequential([
    Dense(64, activation='relu', input_shape=(X_poly_train_scaled.shape[1],)),
    Dense(64, activation='relu'),
    Dense(1)
])

# Compile the model
model_tf.compile(optimizer='adam', loss='mean_squared_error', metrics=['mse'])

# Train the model
history_tf = model_tf.fit(X_poly_train_scaled, y_poly_train, epochs=100, batch_size=32, validation_split=0.2)

# Evaluate the model
mse_tf, _ = model_tf.evaluate(X_poly_test_scaled, y_poly_test)
print("Mean Squared Error (TensorFlow/Keras):", mse_tf)
print("-----------------------------------------------------------------------")

# Implement classification
# Define the model for classification for TenserFlow
model_tf_classification = Sequential([
    Dense(64, activation='relu', input_shape=(X_class_train_scaled.shape[1],)),
    Dense(64, activation='relu'),
    Dense(1, activation='sigmoid')
])

# Compile the model
model_tf_classification.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

# Train the model
history_tf_classification = model_tf_classification.fit(X_class_train_scaled, y_class_train, epochs=100, batch_size=32, validation_split=0.2)

# Evaluate the model
loss_tf_classification, accuracy_tf_classification = model_tf_classification.evaluate(X_class_test_scaled, y_class_test)
print("Test Loss (TensorFlow/Keras):", loss_tf_classification)
print("Test Accuracy (TensorFlow/Keras):", accuracy_tf_classification)


Epoch 1/100
3/3 [==============================] - 2s 231ms/step - loss: 17.6455 - mse: 17.6455 - val_loss: 12.9945 - val_mse: 12.9945
Epoch 2/100
3/3 [==============================] - 0s 112ms/step - loss: 15.9432 - mse: 15.9432 - val_loss: 11.5183 - val_mse: 11.5183
Epoch 3/100
3/3 [==============================] - 0s 80ms/step - loss: 14.4020 - mse: 14.4020 - val_loss: 10.1804 - val_mse: 10.1804
Epoch 4/100
3/3 [==============================] - 0s 46ms/step - loss: 12.9730 - mse: 12.9730 - val_loss: 8.9353 - val_mse: 8.9353
Epoch 5/100
3/3 [==============================] - 0s 24ms/step - loss: 11.6180 - mse: 11.6180 - val_loss: 7.7660 - val_mse: 7.7660
Epoch 6/100
3/3 [==============================] - 0s 16ms/step - loss: 10.2894 - mse: 10.2894 - val_loss: 6.6676 - val_mse: 6.6676
Epoch 7/100
3/3 [==============================] - 0s 15ms/step - loss: 9.0490 - mse: 9.0490 - val_loss: 5.6466 - val_mse: 5.6466
Epoch 8/100
3/3 [==============================] - 0s 15ms/step - loss

In [8]:
import torch
import torch.nn as nn
import torch.optim as optim
from torch.utils.data import TensorDataset, DataLoader

# Define the model
class PolyRegressionModel(nn.Module):
    def __init__(self, input_dim):
        super(PolyRegressionModel, self).__init__()
        self.fc1 = nn.Linear(input_dim, 64)
        self.fc2 = nn.Linear(64, 64)
        self.fc3 = nn.Linear(64, 1)

    def forward(self, x):
        x = torch.relu(self.fc1(x))
        x = torch.relu(self.fc2(x))
        x = self.fc3(x)
        return x

# Convert numpy arrays to PyTorch tensors
X_poly_train_tensor = torch.tensor(X_poly_train_scaled, dtype=torch.float32)
y_poly_train_tensor = torch.tensor(y_poly_train.reshape(-1, 1), dtype=torch.float32)

X_poly_test_tensor = torch.tensor(X_poly_test_scaled, dtype=torch.float32)
y_poly_test_tensor = torch.tensor(y_poly_test.reshape(-1, 1), dtype=torch.float32)

# Create DataLoader
train_dataset = TensorDataset(X_poly_train_tensor, y_poly_train_tensor)
train_loader = DataLoader(train_dataset, batch_size=32, shuffle=True)

# Initialize the model
model_pytorch = PolyRegressionModel(X_poly_train_scaled.shape[1])

# Define loss function and optimizer
criterion = nn.MSELoss()
optimizer = optim.Adam(model_pytorch.parameters(), lr=0.001)

# Train the model
for epoch in range(100):
    running_loss = 0.0
    for inputs, labels in train_loader:
        optimizer.zero_grad()
        outputs = model_pytorch(inputs)
        loss = criterion(outputs, labels)
        loss.backward()
        optimizer.step()
        running_loss += loss.item()
    print(f"Epoch {epoch+1}, Loss: {running_loss}")

# Evaluate the model
with torch.no_grad():
    outputs = model_pytorch(X_poly_test_tensor)
    mse_pytorch = criterion(outputs, y_poly_test_tensor)
    print("Mean Squared Error (PyTorch):", mse_pytorch.item())
print("-----------------------------------------------------------------------")

# Implement classification
# Define the model for classification in PyTorch
class ClassificationModel(nn.Module):
    def __init__(self, input_dim):
        super(ClassificationModel, self).__init__()
        self.fc1 = nn.Linear(input_dim, 64)
        self.fc2 = nn.Linear(64, 64)
        self.fc3 = nn.Linear(64, 1)

    def forward(self, x):
        x = torch.relu(self.fc1(x))
        x = torch.relu(self.fc2(x))
        x = torch.sigmoid(self.fc3(x))
        return x


# Initialize the model
model_pytorch_classification = ClassificationModel(X_class_train_scaled.shape[1])

# Convert numpy arrays to PyTorch tensors for classification
X_class_train_tensor = torch.tensor(X_class_train_scaled, dtype=torch.float32)
y_class_train_tensor = torch.tensor(y_class_train, dtype=torch.float32).reshape(-1, 1)

X_class_test_tensor = torch.tensor(X_class_test_scaled, dtype=torch.float32)
y_class_test_tensor = torch.tensor(y_class_test, dtype=torch.float32).reshape(-1, 1)

# Create DataLoader for classification
train_data_classification = TensorDataset(X_class_train_tensor, y_class_train_tensor)
train_loader_classification = DataLoader(train_data_classification, batch_size=32, shuffle=True)

test_data_classification = TensorDataset(X_class_test_tensor, y_class_test_tensor)
test_loader_classification = DataLoader(test_data_classification, batch_size=32, shuffle=False)

# Define loss function and optimizer
criterion_classification = nn.BCELoss()
optimizer_classification = optim.Adam(model_pytorch_classification.parameters(), lr=0.001)

# Train the model
for epoch in range(100):
    running_loss = 0.0
    for inputs, labels in train_loader_classification:
        optimizer_classification.zero_grad()
        outputs = model_pytorch_classification(inputs)
        loss_classification = criterion_classification(outputs, labels)
        loss_classification.backward()
        optimizer_classification.step()
        running_loss += loss_classification.item()
    print(f"Epoch {epoch+1}, Loss: {running_loss}")

# Evaluate the model
with torch.no_grad():
    outputs_classification = model_pytorch_classification(X_class_test_tensor)
    loss_classification = criterion_classification(outputs_classification, y_class_test_tensor)
    predictions = (outputs_classification > 0.5).float()
    accuracy_pytorch_classification = torch.mean((predictions == y_class_test_tensor).float())
    print("Test Loss (PyTorch):", loss_classification.item())
    print("Test Accuracy (PyTorch):", accuracy_pytorch_classification.item())


Epoch 1, Loss: 50.64205455780029
Epoch 2, Loss: 46.37148857116699
Epoch 3, Loss: 43.307146072387695
Epoch 4, Loss: 40.848365783691406
Epoch 5, Loss: 37.393853187561035
Epoch 6, Loss: 34.89534282684326
Epoch 7, Loss: 30.88826274871826
Epoch 8, Loss: 28.516469478607178
Epoch 9, Loss: 24.55151653289795
Epoch 10, Loss: 21.053458213806152
Epoch 11, Loss: 17.380541801452637
Epoch 12, Loss: 14.035048007965088
Epoch 13, Loss: 11.5358145236969
Epoch 14, Loss: 8.92451524734497
Epoch 15, Loss: 7.464125394821167
Epoch 16, Loss: 6.47645890712738
Epoch 17, Loss: 5.9928401708602905
Epoch 18, Loss: 5.293258428573608
Epoch 19, Loss: 5.852438926696777
Epoch 20, Loss: 5.865965247154236
Epoch 21, Loss: 5.552149474620819
Epoch 22, Loss: 5.563105821609497
Epoch 23, Loss: 5.352930426597595
Epoch 24, Loss: 5.192534565925598
Epoch 25, Loss: 5.061388969421387
Epoch 26, Loss: 5.036796927452087
Epoch 27, Loss: 4.667187690734863
Epoch 28, Loss: 4.7242395877838135
Epoch 29, Loss: 4.399576961994171
Epoch 30, Loss: 4

---

#Conclusion from training two frameworks on the provided dataset

---



##Data compare: TenserFlow

```
Epoch 1/100
3/3 [==============================] - 2s 231ms/step - loss: 17.6455 - mse: 17.6455 - val_loss: 12.9945 - val_mse: 12.9945
...
Epoch 100/100
3/3 [==============================] - 0s 15ms/step - loss: 1.0788 - mse: 1.0788 - val_loss: 2.1229 - val_mse: 2.1229
1/1 [==============================] - 0s 23ms/step - loss: 2.5335 - mse: 2.5335
Mean Squared Error (TensorFlow/Keras): 2.5335023403167725
-----------------------------------------------------------------------
Epoch 1/100
3/3 [==============================] - 1s 91ms/step - loss: 0.7682 - accuracy: 0.1765 - val_loss: 0.7368 - val_accuracy: 0.2941
...
Epoch 100/100
3/3 [==============================] - 0s 19ms/step - loss: 0.1463 - accuracy: 0.9265 - val_loss: 0.1271 - val_accuracy: 1.0000
1/1 [==============================] - 0s 34ms/step - loss: 0.5917 - accuracy: 0.8636
Test Loss (TensorFlow/Keras): 0.5916541218757629
Test Accuracy (TensorFlow/Keras): 0.8636363744735718

```



##Data compare: PyTorch



```
Epoch 1, Loss: 50.64205455780029
...
Epoch 100, Loss: 3.4720852971076965
Mean Squared Error (PyTorch): 1.9616214036941528
-----------------------------------------------------------------------
Epoch 1, Loss: 1.9761527180671692
...
Epoch 100, Loss: 0.37988586723804474
Test Loss (PyTorch): 0.5991433262825012
Test Accuracy (PyTorch): 0.8636363744735718
```



###Based on the results provided, both frameworks, PyTorch and TensorFlow, were used to train and evaluate a deep neural network model. The training process involved multiple epochs, with the loss decreasing over time, indicating that the model was learning from the data. In the case of PyTorch, the mean squared error decreased from approximately 50.64 to 3.47, while for TensorFlow, the loss decreased from around 17.6455 to 1.0788. Additionally, the test accuracy for both PyTorch and TensorFlow was reported to be approximately 86.36%, indicating that the models performed well on unseen data. These results suggest that both frameworks were effective in training the model, but PyTorch showcased slightly better performance in terms of convergence speed.